qutip/qip/compiler/gatecompiler.py

```diff
@@ -95,23 +123,169 @@ def decompose(self, gates):
-        self.coeff_list = []
+        if isinstance(circuit, QubitCircuit):
+            gates = circuit.gates
+        else:
+            gates = circuit
+        if args is not None:
+            self.args.update(args)
+        instruction_list = []
         for gate in gates:
-            if gate.name not in self.gate_decomps:
+            if gate.name not in self.gate_compiler:
                 raise ValueError("Unsupported gate %s" % gate.name)
-            self.gate_decomps[gate.name](gate)
-        coeffs = np.vstack(self.coeff_list).T
-
-        tlist = np.empty(len(self.dt_list))
-        t = 0
-        for i in range(len(self.dt_list)):
-            t += self.dt_list[i]
-            tlist[i] = t
-        return np.hstack([[0], tlist]), coeffs
+            instruction = self.gate_compiler[gate.name](gate, self.args)
+            if instruction is None:
+                continue  # neglecting global phase gate
+            instruction_list += instruction
+        if not instruction_list:
+            return None, None
+        if self.pulse_dict is not None:
+            num_controls = len(self.pulse_dict)
+        else:  # if pulse_dict is not given, compute the number of pulses
+            num_controls = 0
+            for instruction in instruction_list:
+                for pulse_index, _ in instruction.pulse_info:
+                    num_controls = max(num_controls, pulse_index)
+            num_controls += 1
+
+        # schedule
+        # scheduled_start_time:
+        #   An ordered list of the start_time for each pulse,
+        #   corresponding to gates in the instruction_list.
+        # instruction_list reordered according to the scheduled result
+        if schedule_mode:
+            scheduler = Scheduler(schedule_mode)
+            scheduled_start_time = scheduler.schedule(instruction_list)
+            time_ordered_pos = np.argsort(scheduled_start_time)
+            instruction_list = [instruction_list[i] for i in time_ordered_pos]
+            scheduled_start_time.sort()
+        else:  # no scheduling
+            time_ordered_pos = list(range(0, len(instruction_list)))
+            scheduled_start_time = [0.]
+            for instruction in instruction_list:
+                scheduled_start_time.append(
+                    instruction.duration + scheduled_start_time[-1])
+            scheduled_start_time = scheduled_start_time[:-1]
+
+        # compile
+        # An instruction can be composed from several different pulse elements.
+        # We separate them an assign them to each pulse index.
+        pulse_instructions = [[] for tmp in range(num_controls)]
+        for instruction, start_time in \
+                zip(instruction_list, scheduled_start_time):
+            for pulse_name, coeff in instruction.pulse_info:
+                if self.pulse_dict is not None:
+                    pulse_ind = self.pulse_dict[pulse_name]
+                else:
+                    pulse_ind = pulse_name
```

Comment:
Indeed this is the difference.

I'm not sure if this is a good design so I'm just writing things down to clarify:

The intention behind `pulse_dict` is to have a map between the string name of a pulse and its location in the list `Processor.pulses`, so that while defining gate compilation, one can use string names instead of numbers. More intuitive. In all predefined `Processor`s, this map is defined in the `Processor` subclass and passed to the compiler when calling it. If `pulse_dict` is not given, it should use the integer index.

That said, I don't see it particular useful here to partially overriding this. Instead, if a `pulse_name` is not found in the dictionary, it is more likely a result of a typo: the name is mistyped. Allow it to go further will probably make it harder to debug: a string is assigned to something called `pulse_ind`.

Reply:
Unifying the type of `pulse_dict` to `dict` does sounds nice. Maybe I can add a type check for `pulse_ind` and get rid of `pulse_dict=None`

Reply:
Sure, if it's got to be either all of them or none of them then this makes sense.  A comment up top wouldn't go amiss - because of Python's weak "truthiness", you'd usually expect `{}` and `None` to behave the same if they're both valid inputs.

I'm not wild about the idea of exposing a public attribute that has variable type (`dict` vs `None`), but it's a minor thing and it's definitely up to you.

Reply:
After a second thought on this. I changed it and use `pulse_dict={}` to replace `pulse_dict=None`. In the end, an empty `pulse_dict` doesn't mean anything useful here. It is going to always through an error. Using it instead of `pulse_dict=None` may save some handling with `None` as you suggested in a few comments above.